### PR TITLE
chore: implement encoding.Unmarshal

### LINKED ISSUE #951: Introduce `encoding.Unmarshal` and replace `json.Unmarshal` where needed.
Messages that are encoded with `encoding.Marshal` cannot be correctly decoded with `json.Unmarshal`

@worstell knew we were gonna need this :)

### LINKED ISSUE #971: Implement `go-runtime/encoding.Unmarshal()`
This should be the mirror of `encoding.Marshal()`.

### PR DESCRIPTION
**Pull request by worstell**
fixes #971
fixes #951

### CODE REVIEW

File: go-runtime/encoding/encoding.go [hunk 1]
```diff
@@ -32,11 +35,11 @@ func encodeValue(v reflect.Value, w *bytes.Buffer) error {
 	}
 	t := v.Type()
 	switch {
-	case t.Kind() == reflect.Ptr && t.Elem().Implements(jsonUnmarshaler):
+	case t.Kind() == reflect.Ptr && t.Elem().Implements(jsonMarshaler):
 		v = v.Elem()
 		fallthrough
 
-	case t.Implements(jsonUnmarshaler):
+	case t.Implements(jsonMarshaler):
```
Comment (alecthomas) [inline go-runtime/encoding/encoding.go:42]: Nice catch 🤦‍♂️

File: go-runtime/encoding/encoding.go [hunk 2]
```diff
@@ -198,3 +201,219 @@ func encodeString(v reflect.Value, w *bytes.Buffer) error {
...
+	}
+
+	for i := 0; i < v.NumField(); i++ {
+		field := v.Field(i)
+		fieldType := v.Type().Field(i)
+		jsonKey := strcase.ToLowerCamel(fieldType.Name)
+		if _, ok := jsonData[jsonKey]; ok {
+			fieldTypeStr := fieldType.Type.String()
+			switch {
+			case fieldTypeStr == "*Unit" || fieldTypeStr == "Unit":
```
Comment (alecthomas) [inline go-runtime/encoding/encoding.go:306]: This could be a direct comparison with `reflect.TypeOf(ftl.Unit{})` et al?

**Review by alecthomas — CHANGES_REQUESTED** [review]
Good start, but this needs some tests.

File: go-runtime/encoding/encoding.go [hunk 3]
```diff
@@ -198,3 +201,219 @@ func encodeString(v reflect.Value, w *bytes.Buffer) error {
...
+	v.SetBool(b)
+	return nil
+}
+
+func decodeBytes(data []byte, v reflect.Value) error {
+	var s string
+	if err := json.Unmarshal(data, &s); err != nil {
+		return err
+	}
+	b, err := base64.StdEncoding.DecodeString(s)
```
Comment (alecthomas) [inline go-runtime/encoding/encoding.go:363]: Can't decode directly into []byte and get this for free?

File: go-runtime/encoding/encoding.go [hunk 4]
```diff
@@ -198,3 +201,219 @@ func encodeString(v reflect.Value, w *bytes.Buffer) error {
...
+
+	return nil
+}
+
+func decodeMap(data []byte, v reflect.Value) error {
+	if v.IsNil() {
+		v.Set(reflect.MakeMap(v.Type()))
+	}
+
+	var intfMap map[string]interface{}
```
Comment (alecthomas) [inline go-runtime/encoding/encoding.go:401]: For modern Go use `interface{}` -> `any`

**Review by alecthomas — COMMENTED** [review]
Would it be better to use the streaming JSON decoder?

File: go-runtime/encoding/encoding.go (hunk 3 above)
Comment (worstell) [reply go-runtime/encoding/encoding.go]: oop yes

File: go-runtime/encoding/encoding.go (hunk 2 above)
Comment (worstell) [reply go-runtime/encoding/encoding.go:313]: this introduces a dependency cycle because `ftl` depends on encoding (it's used in `option.go`)

**Comment by worstell** [comment]
> Would it be better to use the streaming JSON decoder?

good idea! updated

**Review by alecthomas — APPROVED** [review]
Nice!

**Pull request merged**